pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,358 @@
+		return nil, nil, err
+	}
+
+	c := Config{}
+	c.Token, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.Token, "CHERRYSERVERS_TOKEN")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"token\" field, error = %v", err)
+	}
+	c.Plan, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Plan)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	c.ProjectID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.ProjectID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"projectID\" field, error = %v", err)
+	}
+
+	c.TeamID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.TeamID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"teamID\" field, error = %v", err)
+	}
+
+	c.Location, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.Location)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get the value of \"location\" field, error = %v", err)
+	}
+	c.Labels = rawConfig.Labels
+	return &c, &pconfig, err
+}
+
+func getNameForOS(os providerconfig.OperatingSystem) (string, error) {
+	switch os {
+	case providerconfig.OperatingSystemUbuntu:
+		return "Ubuntu 18.04 64bit", nil
+	case providerconfig.OperatingSystemCentOS:
+		return "CentOS 7 64bit", nil
+	}
+	return "", providerconfig.ErrOSNotSupported
+}
+
+func getClient(token string) *cherrygo.Client {
+	httpClient := &http.Client{}
+	return cherrygo.NewClientWithAuthVar(httpClient, token)
+}
+
+func getTeamByID(client *cherrygo.Client, ID string) *cherrygo.Teams {
+	teams, _, err := client.Teams.List()
+	if err != nil {
+		return nil
+	}
+
+	for _, team := range teams {
+		if string(team.ID) == ID {
+			return &team
+		}
+	}
+	return nil
+}
+
+func getPlanByName(client *cherrygo.Client, planName string, teamID string) *cherrygo.Plans {
+	tid, err := strconv.Atoi(teamID)
+	if err != nil {
+		return nil
+	}
+
+	plans, _, err := client.Plans.List(tid)
+	if err != nil {
+		return nil
+	}
+
+	for _, plan := range plans {
+		if plan.Name == planName {
+			return &plan
+		}
+	}
+	return nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.Token == "" {
+		return errors.New("token is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.Token)
+
+	if c.TeamID != "" {
+		if getTeamByID(client, c.TeamID) == nil {
+			return fmt.Errorf("failed to get teamID")
+		}
+	}
+
+	if c.Plan != "" {
+		if getPlanByName(client, c.Plan, c.TeamID) == nil {
+			return fmt.Errorf("failed to get plan")
+		}
+	}
+
+	if c.Location != "" {
+		if c.Location != "EU-East-1" && c.Location != "EU-West-1" {
```

Comment:
Is there a more dynamic way of validating this? For example, if there is an API endpoint, we can use that one. If not, this way should work too.

Reply:
Shouldn't this be or (`||`) instead of and (`&&`)?

Reply:
No API endpoints for regions available here. I will ask cherry API developer to add it.